syntax = "proto2";
package oneflow;

// The order of collective boxing task nodes are partually decided by ctrl edges on all ranks.
// We must gather all ctrl edges to master. make sure master makes the right decision (no deadlock).
// Only used in rank compiler.
message PortableCtrlNode {
  oneof ctrl_node_type {
    int64 transport_task_id = 1; // all transport_task_ids are shared on all ranks.
    string compute_task_op_name = 2; // all compute_task_op_names are known by master.
  }
}

message PortableCtrlEdge {
  required PortableCtrlNode src = 1;
  required PortableCtrlNode dst = 2;
}
